Resteasy Reactive report NoSuchMethod Exception with quarkus-keycloak-admin-client dependencies in native mode

**Describe the bug**
(Describe the problem clearly and concisely.)

I switch my project from Resteasy to Resteasy Reactive recently. When I call post or put endpiont  in native mode., NoSuchMethodException will be report.

It  work fine in JVM mode.

```log
2021-03-12 22:46:11,897 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (executor-thread-1) Request failed: java.lang.RuntimeException: java.lang.NoSuchMethodException: org.acme.CategoryResource.add(org.acme.Category)
        at org.jboss.resteasy.reactive.server.spi.ResteasyReactiveResourceInfo.getMethod(ResteasyReactiveResourceInfo.java:55)
        at org.jboss.resteasy.reactive.server.spi.ResteasyReactiveResourceInfo.getParameterAnnotations(ResteasyReactiveResourceInfo.java:89)
        at org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.getAnnotations(RequestDeserializeHandler.java:115)
        at org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.isReadable(RequestDeserializeHandler.java:101)
        at org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.handle(RequestDeserializeHandler.java:62)
        at org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.handle(RequestDeserializeHandler.java:25)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:117)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2415)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at java.lang.Thread.run(Thread.java:834)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:519)
        at com.oracle.svm.core.windows.WindowsJavaThreads.osThreadStartRoutine(WindowsJavaThreads.java:138)
Caused by: java.lang.NoSuchMethodException: org.acme.CategoryResource.add(org.acme.Category)
        at java.lang.Class.getMethod(DynamicHub.java:1089)
        at org.jboss.resteasy.reactive.server.spi.ResteasyReactiveResourceInfo.getMethod(ResteasyReactiveResourceInfo.java:50)
        ... 14 more
```

My project dependent  quarkus-keycloak-admin-client.
```xml
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-resteasy-reactive-jsonb</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-resteasy-reactive</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-keycloak-admin-client</artifactId>
        </dependency>
```
When  I remove the dependency on quarkus-keycloak-admin-client, the project will work fine. So I think the problem has something to do with  quarkus-keycloak-admin-client.


**To Reproduce**

Link to a small reproducer (preferably a Maven project if the issue is not Gradle-specific).

Or attach an archive containing the reproducer to the issue.

[demo1.zip](https://github.com/quarkusio/quarkus/files/6130852/demo1.zip)

Steps to reproduce the behavior:
1.   mvn package -Pnative
2.   run  program
3.  call post endpoint 

**Configuration**
```properties
# Add your application.properties here, if applicable.

```

**Screenshots**
(If applicable, add screenshots to help explain your problem.)
![Category](https://user-images.githubusercontent.com/48646226/110959416-a14e8280-8388-11eb-934c-eff4a5b66d65.png)

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
 - Output of `java -version`:   build 11.0.7+8-LTS
 - GraalVM version (if different from Java):   graalvm-ce-java11-21.0.0.2
 - Quarkus version or git rev:    1.12.2
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`):  3.6.3
 - OS: windows 10  or  Cent OS 7.5
   
**Additional context**
(Add any other context about the problem here.)

--- a/extensions/resteasy-reactive/quarkus-jaxrs-client/deployment/src/main/java/io/quarkus/resteasy/reactive/client/deployment/JaxrsClientProcessor.java
+++ b/extensions/resteasy-reactive/quarkus-jaxrs-client/deployment/src/main/java/io/quarkus/resteasy/reactive/client/deployment/JaxrsClientProcessor.java
@@ -96,7 +96,9 @@ import io.quarkus.resteasy.reactive.common.deployment.ResourceScanningResultBuil
 import io.quarkus.resteasy.reactive.common.deployment.SerializersUtil;
 import io.quarkus.resteasy.reactive.common.runtime.ResteasyReactiveConfig;
 import io.quarkus.resteasy.reactive.spi.MessageBodyReaderBuildItem;
+import io.quarkus.resteasy.reactive.spi.MessageBodyReaderOverrideBuildItem;
 import io.quarkus.resteasy.reactive.spi.MessageBodyWriterBuildItem;
+import io.quarkus.resteasy.reactive.spi.MessageBodyWriterOverrideBuildItem;
 import io.quarkus.runtime.RuntimeValue;
 
 public class JaxrsClientProcessor {
@@ -132,6 +134,8 @@ public class JaxrsClientProcessor {
             BuildProducer<ReflectiveClassBuildItem> reflectiveClassBuildItemBuildProducer,
             List<MessageBodyReaderBuildItem> messageBodyReaderBuildItems,
             List<MessageBodyWriterBuildItem> messageBodyWriterBuildItems,
+            List<MessageBodyReaderOverrideBuildItem> messageBodyReaderOverrideBuildItems,
+            List<MessageBodyWriterOverrideBuildItem> messageBodyWriterOverrideBuildItems,
             List<JaxrsClientEnricherBuildItem> enricherBuildItems,
             BeanArchiveIndexBuildItem beanArchiveIndexBuildItem,
             Optional<ResourceScanningResultBuildItem> resourceScanningResultBuildItem,
@@ -147,7 +151,8 @@ public class JaxrsClientProcessor {
         Serialisers serialisers = recorder.createSerializers();
 
         SerializersUtil.setupSerializers(recorder, reflectiveClassBuildItemBuildProducer, messageBodyReaderBuildItems,
-                messageBodyWriterBuildItems, beanContainerBuildItem, applicationResultBuildItem, serialisers,
+                messageBodyWriterBuildItems, messageBodyReaderOverrideBuildItems, messageBodyWriterOverrideBuildItems,
+                beanContainerBuildItem, applicationResultBuildItem, serialisers,
                 RuntimeType.CLIENT);
 
         if (!resourceScanningResultBuildItem.isPresent()

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/deployment/src/main/java/io/quarkus/resteasy/reactive/common/deployment/ResteasyReactiveCommonProcessor.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/deployment/src/main/java/io/quarkus/resteasy/reactive/common/deployment/ResteasyReactiveCommonProcessor.java
@@ -10,6 +10,7 @@ import java.util.Set;
 import java.util.function.Consumer;
 import java.util.stream.Collectors;
 
+import javax.ws.rs.Priorities;
 import javax.ws.rs.RuntimeType;
 import javax.ws.rs.ext.RuntimeDelegate;
 
@@ -249,8 +250,13 @@ public class ResteasyReactiveCommonProcessor {
                 if (constrainedToInstance != null) {
                     runtimeType = RuntimeType.valueOf(constrainedToInstance.value().asEnum());
                 }
+                int priority = Priorities.USER;
+                AnnotationInstance priorityInstance = writerClass.classAnnotation(ResteasyReactiveDotNames.PRIORITY);
+                if (priorityInstance != null) {
+                    priority = priorityInstance.value().asInt();
+                }
                 messageBodyWriterBuildItemBuildProducer.produce(new MessageBodyWriterBuildItem(writerClassName,
-                        typeParameters.get(0).name().toString(), mediaTypeStrings, runtimeType, false));
+                        typeParameters.get(0).name().toString(), mediaTypeStrings, runtimeType, false, priority));
             }
         }
 
@@ -274,8 +280,13 @@ public class ResteasyReactiveCommonProcessor {
                 if (constrainedToInstance != null) {
                     runtimeType = RuntimeType.valueOf(constrainedToInstance.value().asEnum());
                 }
+                int priority = Priorities.USER;
+                AnnotationInstance priorityInstance = readerClass.classAnnotation(ResteasyReactiveDotNames.PRIORITY);
+                if (priorityInstance != null) {
+                    priority = priorityInstance.value().asInt();
+                }
                 messageBodyReaderBuildItemBuildProducer.produce(new MessageBodyReaderBuildItem(readerClassName,
-                        typeParameters.get(0).name().toString(), mediaTypeStrings, runtimeType, false));
+                        typeParameters.get(0).name().toString(), mediaTypeStrings, runtimeType, false, priority));
                 reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, false, readerClassName));
             }
         }

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/deployment/src/main/java/io/quarkus/resteasy/reactive/common/deployment/ResteasyReactiveCommonProcessor.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/deployment/src/main/java/io/quarkus/resteasy/reactive/common/deployment/ResteasyReactiveCommonProcessor.java
@@ -49,13 +49,18 @@ import io.quarkus.resteasy.reactive.spi.AbstractInterceptorBuildItem;
 import io.quarkus.resteasy.reactive.spi.ContainerRequestFilterBuildItem;
 import io.quarkus.resteasy.reactive.spi.ContainerResponseFilterBuildItem;
 import io.quarkus.resteasy.reactive.spi.MessageBodyReaderBuildItem;
+import io.quarkus.resteasy.reactive.spi.MessageBodyReaderOverrideBuildItem;
 import io.quarkus.resteasy.reactive.spi.MessageBodyWriterBuildItem;
+import io.quarkus.resteasy.reactive.spi.MessageBodyWriterOverrideBuildItem;
 import io.quarkus.resteasy.reactive.spi.ReaderInterceptorBuildItem;
 import io.quarkus.resteasy.reactive.spi.WriterInterceptorBuildItem;
 import io.quarkus.security.spi.AdditionalSecuredClassesBuildIem;
 import io.quarkus.security.spi.SecurityTransformerUtils;
 
 public class ResteasyReactiveCommonProcessor {
+
+    private static final int LEGACY_READER_PRIORITY = Priorities.USER * 2; // readers are compared by decreased priority
+    private static final int LEGACY_WRITER_PRIORITY = Priorities.USER / 2; // writers are compared by increased priority
 
     @BuildStep
     void setUpDenyAllJaxRs(CombinedIndexBuildItem index,
@@ -296,6 +301,31 @@ public class ResteasyReactiveCommonProcessor {
     void registerRuntimeDelegateImpl(BuildProducer<ServiceProviderBuildItem> serviceProviders) {
         serviceProviders.produce(new ServiceProviderBuildItem(RuntimeDelegate.class.getName(),
                 RuntimeDelegateImpl.class.getName()));
+    }
+
+    /*
+     * There are some MessageBodyReaders and MessageBodyWriters that are brought in transitively
+     * by the inclusion of the extension like the 'quarkus-keycloak-admin-client'.
+     * We need to make sure that these providers are not selected over the ones that our Quarkus extensions provide.
+     * To do that, we first need to make them built-in (as the spec mandates that non-build-in providers are choosen
+     * over built-in ones) and then we also need to change their priority
+     */
+    @BuildStep
+    void deprioritizeLegacyProviders(BuildProducer<MessageBodyReaderOverrideBuildItem> readers,
+            BuildProducer<MessageBodyWriterOverrideBuildItem> writers) {
+        readers.produce(new MessageBodyReaderOverrideBuildItem(
+                "org.jboss.resteasy.plugins.providers.jackson.ResteasyJackson2Provider", LEGACY_READER_PRIORITY, true));
+        readers.produce(new MessageBodyReaderOverrideBuildItem("com.fasterxml.jackson.jaxrs.json.JacksonJsonProvider",
+                LEGACY_READER_PRIORITY, true));
+        readers.produce(new MessageBodyReaderOverrideBuildItem("com.fasterxml.jackson.jaxrs.json.JacksonJaxbJsonProvider",
+                LEGACY_READER_PRIORITY, true));
+
+        writers.produce(new MessageBodyWriterOverrideBuildItem(
+                "org.jboss.resteasy.plugins.providers.jackson.ResteasyJackson2Provider", LEGACY_WRITER_PRIORITY, true));
+        writers.produce(new MessageBodyWriterOverrideBuildItem("com.fasterxml.jackson.jaxrs.json.JacksonJsonProvider",
+                LEGACY_WRITER_PRIORITY, true));
+        writers.produce(new MessageBodyWriterOverrideBuildItem("com.fasterxml.jackson.jaxrs.json.JacksonJaxbJsonProvider",
+                LEGACY_WRITER_PRIORITY, true));
     }
 
     /**

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/deployment/src/main/java/io/quarkus/resteasy/reactive/common/deployment/SerializersUtil.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/deployment/src/main/java/io/quarkus/resteasy/reactive/common/deployment/SerializersUtil.java
@@ -1,6 +1,8 @@
 package io.quarkus.resteasy.reactive.common.deployment;
 
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 
 import javax.ws.rs.RuntimeType;
 
@@ -13,7 +15,10 @@ import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
 import io.quarkus.resteasy.reactive.common.runtime.ResteasyReactiveCommonRecorder;
 import io.quarkus.resteasy.reactive.spi.MessageBodyReaderBuildItem;
+import io.quarkus.resteasy.reactive.spi.MessageBodyReaderOverrideBuildItem;
+import io.quarkus.resteasy.reactive.spi.MessageBodyReaderWriterOverrideData;
 import io.quarkus.resteasy.reactive.spi.MessageBodyWriterBuildItem;
+import io.quarkus.resteasy.reactive.spi.MessageBodyWriterOverrideBuildItem;
 import io.quarkus.resteasy.reactive.spi.RuntimeTypeItem;
 
 public class SerializersUtil {
@@ -22,15 +27,26 @@ public class SerializersUtil {
             BuildProducer<ReflectiveClassBuildItem> reflectiveClass,
             List<MessageBodyReaderBuildItem> messageBodyReaderBuildItems,
             List<MessageBodyWriterBuildItem> messageBodyWriterBuildItems,
+            List<MessageBodyReaderOverrideBuildItem> messageBodyReaderOverrideBuildItems,
+            List<MessageBodyWriterOverrideBuildItem> messageBodyWriterOverrideBuildItems,
             BeanContainerBuildItem beanContainerBuildItem,
             ApplicationResultBuildItem applicationResultBuildItem,
             Serialisers serialisers, RuntimeType runtimeType) {
 
+        Map<String, MessageBodyReaderWriterOverrideData> writerOverrides = new HashMap<>();
+        for (MessageBodyWriterOverrideBuildItem writerOverride : messageBodyWriterOverrideBuildItems) {
+            writerOverrides.put(writerOverride.getClassName(), writerOverride.getOverrideData());
+        }
         for (MessageBodyWriterBuildItem additionalWriter : RuntimeTypeItem.filter(messageBodyWriterBuildItems,
                 runtimeType)) {
             ResourceWriter writer = new ResourceWriter();
-            writer.setBuiltin(additionalWriter.isBuiltin());
             String writerClassName = additionalWriter.getClassName();
+            MessageBodyReaderWriterOverrideData overrideData = writerOverrides.get(writerClassName);
+            if (overrideData != null) {
+                writer.setBuiltin(overrideData.isBuiltIn());
+            } else {
+                writer.setBuiltin(additionalWriter.isBuiltin());
+            }
             writer.setFactory(FactoryUtils.factory(writerClassName,
                     applicationResultBuildItem.getResult().getSingletonClasses(), recorder,
                     beanContainerBuildItem));
@@ -38,15 +54,29 @@ public class SerializersUtil {
             if (!additionalWriter.getMediaTypeStrings().isEmpty()) {
                 writer.setMediaTypeStrings(additionalWriter.getMediaTypeStrings());
             }
-            writer.setPriority(additionalWriter.getPriority());
+            if (overrideData != null) {
+                writer.setPriority(overrideData.getPriority());
+            } else {
+                writer.setPriority(additionalWriter.getPriority());
+            }
             recorder.registerWriter(serialisers, additionalWriter.getHandledClassName(), writer);
             reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, false, writerClassName));
+        }
+
+        Map<String, MessageBodyReaderWriterOverrideData> readerOverrides = new HashMap<>();
+        for (MessageBodyReaderOverrideBuildItem readerOverride : messageBodyReaderOverrideBuildItems) {
+            readerOverrides.put(readerOverride.getClassName(), readerOverride.getOverrideData());
         }
         for (MessageBodyReaderBuildItem additionalReader : RuntimeTypeItem.filter(messageBodyReaderBuildItems,
                 runtimeType)) {
             ResourceReader reader = new ResourceReader();
-            reader.setBuiltin(additionalReader.isBuiltin());
             String readerClassName = additionalReader.getClassName();
+            MessageBodyReaderWriterOverrideData overrideData = readerOverrides.get(readerClassName);
+            if (overrideData != null) {
+                reader.setBuiltin(overrideData.isBuiltIn());
+            } else {
+                reader.setBuiltin(additionalReader.isBuiltin());
+            }
             reader.setFactory(FactoryUtils.factory(readerClassName,
                     applicationResultBuildItem.getResult().getSingletonClasses(), recorder,
                     beanContainerBuildItem));
@@ -54,7 +84,11 @@ public class SerializersUtil {
             if (!additionalReader.getMediaTypeStrings().isEmpty()) {
                 reader.setMediaTypeStrings(additionalReader.getMediaTypeStrings());
             }
-            reader.setPriority(additionalReader.getPriority());
+            if (overrideData != null) {
+                reader.setPriority(overrideData.getPriority());
+            } else {
+                reader.setPriority(additionalReader.getPriority());
+            }
             recorder.registerReader(serialisers, additionalReader.getHandledClassName(), reader);
             reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, false, readerClassName));
         }

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/deployment/src/main/java/io/quarkus/resteasy/reactive/common/deployment/SerializersUtil.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/deployment/src/main/java/io/quarkus/resteasy/reactive/common/deployment/SerializersUtil.java
@@ -38,6 +38,7 @@ public class SerializersUtil {
             if (!additionalWriter.getMediaTypeStrings().isEmpty()) {
                 writer.setMediaTypeStrings(additionalWriter.getMediaTypeStrings());
             }
+            writer.setPriority(additionalWriter.getPriority());
             recorder.registerWriter(serialisers, additionalWriter.getHandledClassName(), writer);
             reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, false, writerClassName));
         }
@@ -53,6 +54,7 @@ public class SerializersUtil {
             if (!additionalReader.getMediaTypeStrings().isEmpty()) {
                 reader.setMediaTypeStrings(additionalReader.getMediaTypeStrings());
             }
+            reader.setPriority(additionalReader.getPriority());
             recorder.registerReader(serialisers, additionalReader.getHandledClassName(), reader);
             reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, false, readerClassName));
         }

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/spi-deployment/src/main/java/io/quarkus/resteasy/reactive/spi/MessageBodyReaderBuildItem.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/spi-deployment/src/main/java/io/quarkus/resteasy/reactive/spi/MessageBodyReaderBuildItem.java
@@ -2,6 +2,7 @@ package io.quarkus.resteasy.reactive.spi;
 
 import java.util.List;
 
+import javax.ws.rs.Priorities;
 import javax.ws.rs.RuntimeType;
 
 import io.quarkus.builder.item.MultiBuildItem;
@@ -13,18 +14,20 @@ public final class MessageBodyReaderBuildItem extends MultiBuildItem implements 
     private final List<String> mediaTypeStrings;
     private final RuntimeType runtimeType;
     private final boolean builtin;
+    private final Integer priority;
 
     public MessageBodyReaderBuildItem(String className, String handledClassName, List<String> mediaTypeStrings) {
-        this(className, handledClassName, mediaTypeStrings, null, true);
+        this(className, handledClassName, mediaTypeStrings, null, true, Priorities.USER);
     }
 
     public MessageBodyReaderBuildItem(String className, String handledClassName, List<String> mediaTypeStrings,
-            RuntimeType runtimeType, boolean builtin) {
+            RuntimeType runtimeType, boolean builtin, Integer priority) {
         this.className = className;
         this.handledClassName = handledClassName;
         this.mediaTypeStrings = mediaTypeStrings;
         this.runtimeType = runtimeType;
         this.builtin = builtin;
+        this.priority = priority;
     }
 
     public String getClassName() {
@@ -46,5 +49,9 @@ public final class MessageBodyReaderBuildItem extends MultiBuildItem implements 
 
     public boolean isBuiltin() {
         return builtin;
+    }
+
+    public Integer getPriority() {
+        return priority;
     }
 }

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/spi-deployment/src/main/java/io/quarkus/resteasy/reactive/spi/MessageBodyReaderOverrideBuildItem.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/spi-deployment/src/main/java/io/quarkus/resteasy/reactive/spi/MessageBodyReaderOverrideBuildItem.java
@@ -1,0 +1,23 @@
+package io.quarkus.resteasy.reactive.spi;
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+public final class MessageBodyReaderOverrideBuildItem extends MultiBuildItem {
+
+    private final String className;
+    private final MessageBodyReaderWriterOverrideData overrideData;
+
+    public MessageBodyReaderOverrideBuildItem(String className, int priority, boolean builtIn) {
+        this.className = className;
+        this.overrideData = new MessageBodyReaderWriterOverrideData(priority, builtIn);
+
+    }
+
+    public String getClassName() {
+        return className;
+    }
+
+    public MessageBodyReaderWriterOverrideData getOverrideData() {
+        return overrideData;
+    }
+}

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/spi-deployment/src/main/java/io/quarkus/resteasy/reactive/spi/MessageBodyReaderWriterOverrideData.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/spi-deployment/src/main/java/io/quarkus/resteasy/reactive/spi/MessageBodyReaderWriterOverrideData.java
@@ -1,0 +1,20 @@
+package io.quarkus.resteasy.reactive.spi;
+
+public final class MessageBodyReaderWriterOverrideData {
+
+    private final int priority;
+    private final boolean builtIn;
+
+    public MessageBodyReaderWriterOverrideData(int priority, boolean builtIn) {
+        this.priority = priority;
+        this.builtIn = builtIn;
+    }
+
+    public int getPriority() {
+        return priority;
+    }
+
+    public boolean isBuiltIn() {
+        return builtIn;
+    }
+}

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/spi-deployment/src/main/java/io/quarkus/resteasy/reactive/spi/MessageBodyWriterBuildItem.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/spi-deployment/src/main/java/io/quarkus/resteasy/reactive/spi/MessageBodyWriterBuildItem.java
@@ -2,6 +2,7 @@ package io.quarkus.resteasy.reactive.spi;
 
 import java.util.List;
 
+import javax.ws.rs.Priorities;
 import javax.ws.rs.RuntimeType;
 
 import io.quarkus.builder.item.MultiBuildItem;
@@ -13,18 +14,20 @@ public final class MessageBodyWriterBuildItem extends MultiBuildItem implements 
     private final List<String> mediaTypeStrings;
     private final RuntimeType runtimeType;
     private final boolean builtin;
+    private final Integer priority;
 
     public MessageBodyWriterBuildItem(String className, String handledClassName, List<String> mediaTypeStrings) {
-        this(className, handledClassName, mediaTypeStrings, null, false);
+        this(className, handledClassName, mediaTypeStrings, null, false, Priorities.USER);
     }
 
     public MessageBodyWriterBuildItem(String className, String handledClassName, List<String> mediaTypeStrings,
-            RuntimeType runtimeType, boolean builtin) {
+            RuntimeType runtimeType, boolean builtin, Integer priority) {
         this.className = className;
         this.handledClassName = handledClassName;
         this.mediaTypeStrings = mediaTypeStrings;
         this.runtimeType = runtimeType;
         this.builtin = builtin;
+        this.priority = priority;
     }
 
     public String getClassName() {
@@ -46,5 +49,9 @@ public final class MessageBodyWriterBuildItem extends MultiBuildItem implements 
 
     public boolean isBuiltin() {
         return builtin;
+    }
+
+    public Integer getPriority() {
+        return priority;
     }
 }

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/spi-deployment/src/main/java/io/quarkus/resteasy/reactive/spi/MessageBodyWriterOverrideBuildItem.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-common/spi-deployment/src/main/java/io/quarkus/resteasy/reactive/spi/MessageBodyWriterOverrideBuildItem.java
@@ -1,0 +1,23 @@
+package io.quarkus.resteasy.reactive.spi;
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+public final class MessageBodyWriterOverrideBuildItem extends MultiBuildItem {
+
+    private final String className;
+    private final MessageBodyReaderWriterOverrideData overrideData;
+
+    public MessageBodyWriterOverrideBuildItem(String className, int priority, boolean builtIn) {
+        this.className = className;
+        this.overrideData = new MessageBodyReaderWriterOverrideData(priority, builtIn);
+
+    }
+
+    public String getClassName() {
+        return className;
+    }
+
+    public MessageBodyReaderWriterOverrideData getOverrideData() {
+        return overrideData;
+    }
+}

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveProcessor.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveProcessor.java
@@ -111,7 +111,9 @@ import io.quarkus.resteasy.reactive.spi.DynamicFeatureBuildItem;
 import io.quarkus.resteasy.reactive.spi.ExceptionMapperBuildItem;
 import io.quarkus.resteasy.reactive.spi.JaxrsFeatureBuildItem;
 import io.quarkus.resteasy.reactive.spi.MessageBodyReaderBuildItem;
+import io.quarkus.resteasy.reactive.spi.MessageBodyReaderOverrideBuildItem;
 import io.quarkus.resteasy.reactive.spi.MessageBodyWriterBuildItem;
+import io.quarkus.resteasy.reactive.spi.MessageBodyWriterOverrideBuildItem;
 import io.quarkus.runtime.RuntimeValue;
 import io.quarkus.security.AuthenticationCompletionException;
 import io.quarkus.security.AuthenticationFailedException;
@@ -233,6 +235,8 @@ public class ResteasyReactiveProcessor {
             List<DynamicFeatureBuildItem> dynamicFeatures,
             List<MessageBodyReaderBuildItem> additionalMessageBodyReaders,
             List<MessageBodyWriterBuildItem> additionalMessageBodyWriters,
+            List<MessageBodyReaderOverrideBuildItem> messageBodyReaderOverrideBuildItems,
+            List<MessageBodyWriterOverrideBuildItem> messageBodyWriterOverrideBuildItems,
             List<JaxrsFeatureBuildItem> features,
             List<ServerDefaultProducesHandlerBuildItem> serverDefaultProducesHandlers,
             Optional<RequestContextFactoryBuildItem> requestContextFactoryBuildItem,
@@ -459,7 +463,8 @@ public class ResteasyReactiveProcessor {
 
             ServerSerialisers serialisers = new ServerSerialisers();
             SerializersUtil.setupSerializers(recorder, reflectiveClass, additionalMessageBodyReaders,
-                    additionalMessageBodyWriters, beanContainerBuildItem, applicationResultBuildItem, serialisers,
+                    additionalMessageBodyWriters, messageBodyReaderOverrideBuildItems, messageBodyWriterOverrideBuildItems,
+                    beanContainerBuildItem, applicationResultBuildItem, serialisers,
                     RuntimeType.SERVER);
             // built-ins
 

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveProcessor.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveProcessor.java
@@ -144,7 +144,8 @@ public class ResteasyReactiveProcessor {
     @BuildStep
     void vertxIntegration(BuildProducer<MessageBodyWriterBuildItem> writerBuildItemBuildProducer) {
         writerBuildItemBuildProducer.produce(new MessageBodyWriterBuildItem(ServerVertxBufferMessageBodyWriter.class.getName(),
-                Buffer.class.getName(), Collections.singletonList(MediaType.WILDCARD), RuntimeType.SERVER, true));
+                Buffer.class.getName(), Collections.singletonList(MediaType.WILDCARD), RuntimeType.SERVER, true,
+                Priorities.USER));
     }
 
     @BuildStep

--- a/independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/model/ResourceReader.java
+++ b/independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/model/ResourceReader.java
@@ -4,6 +4,7 @@ import java.util.ArrayList;
 import java.util.Collections;
 import java.util.Comparator;
 import java.util.List;
+import javax.ws.rs.Priorities;
 import javax.ws.rs.RuntimeType;
 import javax.ws.rs.core.MediaType;
 import javax.ws.rs.ext.MessageBodyReader;
@@ -16,6 +17,7 @@ public class ResourceReader {
     private List<String> mediaTypeStrings = new ArrayList<>();
     private RuntimeType constraint;
     private boolean builtin = true;
+    private Integer priority = Priorities.USER;
     private volatile List<MediaType> mediaTypes;
     private volatile MessageBodyReader<?> instance;
 
@@ -51,6 +53,14 @@ public class ResourceReader {
 
     public void setBuiltin(boolean builtin) {
         this.builtin = builtin;
+    }
+
+    public Integer getPriority() {
+        return priority;
+    }
+
+    public void setPriority(Integer priority) {
+        this.priority = priority;
     }
 
     public MessageBodyReader<?> instance() {
@@ -92,13 +102,12 @@ public class ResourceReader {
     /**
      * The comparison for now is simple:
      * 1) Application provided writers come first
-     * 2) Then the more specific the media type, the higher the priority
-     * 3) Finally we compare the number of media types
+     * 2) Readers with lower priority come first (same as reader interceptors)
+     * 3) Then the more specific the media type, the higher the priority
+     * 4) Finally we compare the number of media types
      *
      * The spec doesn't seem to mention this sorting being explicitly needed, but there are tests
      * in the TCK that only pass reliably if the Readers are sorted like this
-     *
-     * TODO: if this actually follows the exact same rules as ResourceWriter, we need to refactor
      */
     public static class ResourceReaderComparator implements Comparator<ResourceReader> {
 
@@ -109,6 +118,11 @@ public class ResourceReader {
             int builtInCompare = Boolean.compare(o1.isBuiltin(), o2.isBuiltin());
             if (builtInCompare != 0) {
                 return builtInCompare;
+            }
+
+            int priorityCompare = Integer.compare(o1.getPriority(), o2.getPriority());
+            if (priorityCompare != 0) {
+                return priorityCompare;
             }
 
             List<MediaType> mediaTypes1 = o1.mediaTypes();
@@ -127,7 +141,7 @@ public class ResourceReader {
                 return mediaTypeCompare;
             }
 
-            // TODO: not sure if this makes sense but was added to make the sorting more deterministic
+            // done to make the sorting result deterministic
             return Integer.compare(mediaTypes1.size(), mediaTypes2.size());
         }
     }

--- a/independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/model/ResourceWriter.java
+++ b/independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/model/ResourceWriter.java
@@ -5,6 +5,7 @@ import java.util.ArrayList;
 import java.util.Collections;
 import java.util.Comparator;
 import java.util.List;
+import javax.ws.rs.Priorities;
 import javax.ws.rs.RuntimeType;
 import javax.ws.rs.core.MediaType;
 import javax.ws.rs.ext.MessageBodyWriter;
@@ -18,6 +19,7 @@ public class ResourceWriter {
     private List<String> mediaTypeStrings = new ArrayList<>();
     private RuntimeType constraint;
     private boolean builtin = true;
+    private Integer priority = Priorities.USER;
     private volatile List<MediaType> mediaTypes;
     private volatile ServerMediaType serverMediaType;
     private volatile MessageBodyWriter<?> instance;
@@ -55,6 +57,14 @@ public class ResourceWriter {
 
     public void setBuiltin(boolean builtin) {
         this.builtin = builtin;
+    }
+
+    public Integer getPriority() {
+        return priority;
+    }
+
+    public void setPriority(Integer priority) {
+        this.priority = priority;
     }
 
     public MessageBodyWriter<?> instance() {
@@ -111,8 +121,9 @@ public class ResourceWriter {
     /**
      * The comparison for now is simple:
      * 1) Application provided writers come first
-     * 2) Then the more specific the media type, the higher the priority
-     * 3) Finally we compare the number of media types
+     * 2) Writers higher priority come first (same as writer interceptors)
+     * 3) Then the more specific the media type, the higher the priority
+     * 4) Finally we compare the number of media types
      */
     public static class ResourceWriterComparator implements Comparator<ResourceWriter> {
 
@@ -123,6 +134,11 @@ public class ResourceWriter {
             int builtInCompare = Boolean.compare(o1.isBuiltin(), o2.isBuiltin());
             if (builtInCompare != 0) {
                 return builtInCompare;
+            }
+
+            int priorityCompare = Integer.compare(o2.getPriority(), o1.getPriority());
+            if (priorityCompare != 0) {
+                return priorityCompare;
             }
 
             List<MediaType> mediaTypes1 = o1.mediaTypes();
@@ -141,7 +157,7 @@ public class ResourceWriter {
                 return mediaTypeCompare;
             }
 
-            // TODO: not sure if this makes sense but was added to make the sorting more deterministic
+            // done to make the sorting result deterministic
             return Integer.compare(mediaTypes1.size(), mediaTypes2.size());
         }
     }

